lib/model/queue.go

```diff
@@ -7,65 +7,108 @@
 package model
 
 import (
+	"encoding/binary"
 	"math/rand"
-	"sort"
 	"time"
 
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/diskoverflow"
 	"github.com/syncthing/syncthing/lib/sync"
 )
 
 type jobQueue struct {
-	progress []string
-	queued   []jobQueueEntry
-	mut      sync.Mutex
-}
-
-type jobQueueEntry struct {
-	name     string
-	size     int64
-	modified time.Time
-}
-
-func newJobQueue() *jobQueue {
-	return &jobQueue{
-		mut: sync.NewMutex(),
+	progress       []string
+	queued         diskoverflow.Sorted
+	broughtToFront []string
+	handledAtFront map[string]struct{}
+	location       string
+	order          config.PullOrder
+	shuffleKeys    map[uint64]struct{}
+	mut            sync.Mutex
+}
+
+func newJobQueue(order config.PullOrder, loc string) *jobQueue {
+	q := &jobQueue{
+		handledAtFront: make(map[string]struct{}),
+		location:       loc,
+		order:          order,
+		mut:            sync.NewMutex(),
 	}
+	q.queued = diskoverflow.NewSorted(loc)
+	if order == config.OrderRandom {
+		q.shuffleKeys = make(map[uint64]struct{})
+	}
+	return q
 }
 
 func (q *jobQueue) Push(file string, size int64, modified time.Time) {
+	var key []byte
+	switch q.order {
+	case config.OrderRandom:
+		var n uint64
+		for {
+			n = rand.Uint64()
+			if _, ok := q.shuffleKeys[n]; !ok {
+				q.shuffleKeys[n] = struct{}{}
+				break
+			}
+		}
+		key = make([]byte, 8)
+		binary.BigEndian.PutUint64(key[:], n)
+	case config.OrderAlphabetic:
+		key = []byte(file)
+	case config.OrderSmallestFirst, config.OrderLargestFirst:
+		key = make([]byte, 8)
+		binary.BigEndian.PutUint64(key[:], uint64(size))
+	case config.OrderOldestFirst, config.OrderNewestFirst:
+		key, _ = modified.MarshalText()
+	}
 	q.mut.Lock()
-	q.queued = append(q.queued, jobQueueEntry{file, size, modified})
+	q.queued.Add(key, &queueValue{file})
 	q.mut.Unlock()
 }
 
 func (q *jobQueue) Pop() (string, bool) {
 	q.mut.Lock()
 	defer q.mut.Unlock()
-
-	if len(q.queued) == 0 {
+	if l := len(q.broughtToFront); l > 0 {
+		f := q.broughtToFront[l-1]
+		q.broughtToFront = q.broughtToFront[:l-1]
```

Comment:
It seems bring to front brings the element to front at index zero, yet pop removes it from the end?